packages/shell-api/src/database.ts

```diff
@@ -68,23 +74,37 @@ type AuthDoc = {
   mechanism?: string;
 };
 
+export type Database<
+  M extends GenericServerSideSchema = GenericServerSideSchema,
+  D extends GenericDatabaseSchema = GenericDatabaseSchema
+> = DatabaseImpl<M, D> & {
+  [k in StringKey<D>]: Collection<M, D, D[k], k>;
+};
+
 @shellApiClassDefault
-export default class Database extends ShellApiWithMongoClass {
-  _mongo: Mongo;
-  _name: string;
-  _collections: Record<string, Collection>;
+export class DatabaseImpl<
+  M extends GenericServerSideSchema = GenericServerSideSchema,
+  D extends GenericDatabaseSchema = GenericDatabaseSchema
+> extends ShellApiWithMongoClass {
+  _mongo: Mongo<M>;
+  _name: StringKey<M>;
+  _collections: Record<string, Collection<M, D>>;
   _session: Session | undefined;
-  _cachedCollectionNames: string[] = [];
+  _cachedCollectionNames: StringKey<D>[] = [];
   _cachedHello: Document | null = null;
 
-  constructor(mongo: Mongo, name: string, session?: Session) {
+  _typeLaunder(): Database<M, D> {
```

Comment:
Is this an idiomatic way for this kind of pattern? At least for me, a clearer naming would be `asDatabaseWithSchema` or something similar. `_typeLaunder` doesn't really convey what's happening here, but similarly to the generics - if it's used often enough elsewhere, happy with keeping it.

Reply:
I honestly don't know - I just kept it because that's how it was in the PoC. In general I just tried to "productionize" the PoC which I took as "try and get the tests to pass" 😆 @addaleax what do you think?